Домашнее задание

xafish opened the pull request:
hw06
JPQL, Spring ORM, DAO на основе Spring ORM + JPA

xafish committed:
Домашнее задание
hw06
JPQL, Spring ORM, DAO на основе Spring ORM + JPA

kataus commented on hw06/src/main/java/ru/otus/spring/domain/Author.java:
```diff
@@ -0,0 +1,27 @@
+@AllArgsConstructor
+@NoArgsConstructor(force = true)
+@Entity
+@Table(name = "authors")
```
Удобнее читать, когда аннотации от одного фреймворка собраны вместе

kataus commented on hw06/src/main/java/ru/otus/spring/domain/Comment.java:
```diff
@@ -0,0 +1,34 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private final Long id;
+
+    @ManyToOne(fetch = FetchType.LAZY)
```
Не зря ManyToOne по умолчанию EAGER. Комментарий без книги мало ценности имеет

kataus commented on hw06/src/main/java/ru/otus/spring/repositories/AuthorRepositoryJpa.java:
```diff
@@ -0,0 +1,23 @@
+import ru.otus.spring.domain.Author;
+
+@AllArgsConstructor
+@Repository
```
Что тут используется от аннотации @Repository? Она несколько больше чем @Component или @Service. Если она появляется - то мы ожидаем использование её особенностей

kataus commented on hw06/src/main/java/ru/otus/spring/repositories/BookRepositoryJpa.java:
```diff
@@ -0,0 +1,56 @@
+            " LEFT JOIN FETCH b.genre ";
+
+    @Override
+    @Transactional
```
Какая польза от транзакции здесь? Ещё и с коммитом

kataus commented on hw06/src/main/java/ru/otus/spring/repositories/BookRepositoryJpa.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Override
+    public Book getById(long id) {
+        var query = em.createQuery(bookQueryStr.concat("where b.id = :id"), Book.class);
```
А чем em.find плох? Такой подход не использует силу кэша, а em.find - позволяет им пользоваться. Да и просто проще

kataus commented on hw06/src/main/java/ru/otus/spring/repositories/BookRepositoryJpa.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Override
+    public void deleteById(long id) {
+        var query = em.createQuery("delete from Book b where b.id = :id");
```
А чем em.remove плох? Такой подход достаточно опасен, потому что работает в обход кэша и сущность может там остаться, а из базы уже удалили

kataus commented on hw06/src/main/java/ru/otus/spring/repositories/BookRepositoryJpa.java:
```diff
@@ -0,0 +1,56 @@
+    }
+    @Override
+    public void updateNameById(Long id, String name) {
+        var query = em.createQuery("update Book b set b.name = :name where b.id = :id");
```
Эту работу стоит делегировать ORM, не надо вручную модифицировать базу

kataus commented on hw06/src/main/java/ru/otus/spring/repositories/BookRepositoryJpa.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Override
+    public void addBook(Book book) {
+        em.merge(book);
```
add - это всё таки не merge

kataus commented on hw06/src/main/java/ru/otus/spring/repositories/CommentRepositoryJpa.java:
```diff
@@ -0,0 +1,49 @@
+    private final EntityManager em;
+    private static String commentQueryStr = "select c from Comment c" +
+            " LEFT JOIN FETCH c.book" +
+            " LEFT JOIN FETCH c.book.author " +
```
второй раз уже приходится вручную писать запрос. Стоит эту работу делегировать Hibernate, для того он и нужен

kataus commented on hw06/src/main/java/ru/otus/spring/repositories/CommentRepositoryJpa.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Override
+    public List<Comment> getCommentsByBookName(String bookName) {
+        var query = em.createQuery(commentQueryStr.concat("where c.book.name = :name"), Comment.class);
```
Давайте попробуем работать со связями через ОРМ. Этот метод удалим и решим задачу в бизнес-слое. При этом связь book->comments должна быть ленивой. Без EAGER и подругзки через графы. Задача - именно ленивую связь разрешить

kataus commented on hw06/src/main/java/ru/otus/spring/service/DataManagerShell.java:
```diff
@@ -0,0 +1,142 @@
+            return "Genre with name \"" + genreName + "\" was not found";
+        }
+
+        Book book = new Book(id, bookName, author, genre);
```
Как много бизнес-логики. а вроде класс UI. Single Responsibility страдает, зацепление растет

xafish committed:
Домашнее задание
hw06
JPQL, Spring ORM, DAO на основе Spring ORM + JPA
Исправления по замечаниям

kataus commented on hw06/src/main/java/ru/otus/spring/repositories/BookRepositoryJpa.java:
```diff
@@ -0,0 +1,59 @@
+        em.remove(getById(id));
+    }
+    @Override
+    public void updateNameById(Long id, String name) {
```
Вот этого вообще не надо делать явно в репозитории, это метод для бизнес-слоя. Тут минимально нужные CRUD методы.

kataus commented on hw06/src/main/java/ru/otus/spring/service/BookLibraryServiceImpl.java:
```diff
@@ -0,0 +1,142 @@
+
+    @Override
+    @Transactional
+    public String deleteBook(@ShellOption({"id", "u"}) Long id) {
```
ShellOption уже не тут применяется

kataus commented on hw06/src/main/java/ru/otus/spring/service/BookLibraryServiceImpl.java:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    @Override
+    @Transactional
```
Хорошо бы поставить readOnly=true, будет легче системе
